### PR TITLE
Report class metrics for object detection dataset

### PR DESCRIPTION
**Pull request by Steven-Yiran**
## Summary

Added the `get_class_metrics()` method in this Pull-Request. The goal is to provide a straightforward way for user to understand the performance of their model in each class. The next step is to add the corresponding plotting function once the method is approved.

```python
# Example code snippet

from cleanlab.object_detection.summary import get_class_metrics

# Setup if needed (e.g., data loading, model instantiation):
labels = EXAMPLE_LABELS
predictions = EXAMPLE_MODEL_PREDICTIONS
class_names = EXAMPLE_CLASSNAMES

# Usage and Output:
class_metrics = get_class_metrics(labels, predictions)
# { 'car' : {'accuracy' : 0.67}, 'hand' : {'accuracy' : 0.35} }
```

## Links to Relevant Issues or Conversations

- This Pull-Request follows from the conversation with @ulya-tkch  in #840

## Reviewer Notes
Here is the sudo-code for `_get_class_confusion_matrix()`:
```python
for sample in auxiliary inputs:
  matched_ground_truth_idxs = set()
  for every prediction in predictions:
    # determine the index and value of the highest IoU value
    # if highest IoU value < 0.5 -> FP case
    # determine the index and label of the ground truth label corresponding to the highest IoU value
    # If this ground truth index in matched_ground_truth_idxs -> FP case
    # If the predicted label matches with the ground truth label -> TP case
        # add idx to matched_ground_truth_idxs
    # otherwise -> FP case
  
  for every label in labels:
    # if label not matched by any prediction -> FN case
```

### CODE REVIEW

**Comment by ulya-tkch** [comment]
Hi Steven! Thank you again for splitting this PR into two and continuing with your great contributions. I am a bit busy at the moment but will take a look at this sometime this week and get back to you with some review :D

**Comment by jwmueller** [comment]
We're waiting for the refactor here:
https://github.com/cleanlab/cleanlab/pull/883

to get merged before we review this PR.

**Comment by Steven-Yiran** [comment]
> to get merged before we review this PR.

Thanks for the update!

**Comment by aditya1503** [comment]
> > to get merged before we review this PR.
> 
> Thanks for the update!

hi @Steven-Yiran Most We had code ready for calculating TP,FP 
https://github.com/cleanlab/cleanlab/blob/09d95dc47ff6188c751bf583682c72ce3639536e/cleanlab/object_detection/filter.py#L330

We extended the code to include FN, I'll be pushing a few changes to your branch which will add the functionality you helped build. Your assistance has been instrumental in improving the functionality of Objectlab.

Thank you for your continued support and collaboration. Much appreciated!

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [inline cleanlab/object_detection/summary.py:459]:
```suggestion
    Returns a confusion matrix dictionary for each class containing True Positive, False Positive, and False Negative detections from the object detection model.
```

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [inline cleanlab/object_detection/summary.py:494]: missing one-line of what this method is supposed to do. This is a public-facing method and needs clear description of its purpose

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [inline cleanlab/object_detection/summary.py:510]: please avoid all abbreviations in docstrings, especially in public-facing docstrings. There is little reason to use abbreviations

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [inline cleanlab/object_detection/summary.py:542]: missing short description of what this method is supposed to do. This is a public-facing method and needs clear description of its purpose

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [inline cleanlab/object_detection/summary.py:499]:
```suggestion
        Refer to documentation for this argument in :py:func:`object_detection.filter.find_label_issues <cleanlab.object_detection.filter.find_label_issues>` for further details.
```

Comment (jwmueller) [reply cleanlab/object_detection/summary.py:499]: never just write find_label_issues in docstrings, there are so many find_label_issues methods across the package. Always explicitly write which one you're referring to

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [reply cleanlab/object_detection/summary.py:542]: you need to make it clear that a high-level this is about evaluating the trained object detection model.

Most of the methods in cleanlab are about assessing the data itself. So when we have methods for assessing models, it needs to be clearly stated this method is for assessing models, not data

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [inline cleanlab/object_detection/summary.py:547]: never just write find_label_issues in docstrings, there are so many find_label_issues methods across the package. Always explicitly write which one you're referring to

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [inline cleanlab/object_detection/summary.py:551]: never just write find_label_issues in docstrings, there are so many find_label_issues methods across the package. Always explicitly write which one you're referring to

File: cleanlab/object_detection/summary.py
Comment (jwmueller) [reply cleanlab/object_detection/summary.py:494]: you need to make it clear that a high-level this is about evaluating the trained object detection model.

Most of the methods in cleanlab are about assessing the data itself. So when we have methods for assessing models, it needs to be clearly stated this method is for assessing models, not data

**Comment by jwmueller** [comment]
Closes https://github.com/cleanlab/cleanlab/issues/809

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:540]: [very nit] If it is not too difficult, can this function also accept `class_names` and that just changes the keys returned in per_class_metrics to specific class names? This happens in all the over visualization/summary functions so it would be nice for consistency.

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:491]: Naming convention for function is strange, is this function public or private? If it is private it should be `_get_average_per_class_confusion_matrix()` if public then `get_average_per_class_confusion_matrix()`and it should contain a detailed docsting summarizing what it does.

Comment (Steven-Yiran) [reply cleanlab/object_detection/summary.py]: I assumed this function is public-facing, so I added the docstring summarization and deleted the trailing `_` in the new commit.

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:487]: I would double check below contains the same logic but to me this method was a bit confusing. Not sure what "x" keys were for counter_dict[i][x] so better to use strings (it seems they were tp,fp,fn returned by _calculate_true_positives_false_positives per image). Also the last for loop essentially copying over the dictionary should be avoided if possible with mypy.

```suggestion
    for class_num in range(num_classes):
        pred_bboxes, lab_bboxes = _filter_by_class(labels, predictions, class_num)
        tpfpfn = pool.starmap(
            _calculate_true_positives_false_positives,
            zip(
                pred_bboxes,
                lab_bboxes,
                [iou_threshold for _ in range(num_images)],
                [True for _ in range(num_images)],
            ),
        )
        for image_idx, tpfpfn_per_image in enumerate(tpfpfn):
            tp, fp, fn = tpfpfn_per_image
            counter_dict[class_num]["TP"] += np.sum(tp)
            counter_dict[class_num]["FP"] += np.sum(fp)
            counter_dict[class_num]["FN"] += np.sum(fn)
    
    results: DefaultDict[int, Dict[str, int]] = collections.defaultdict(dict)
    results = {k: dict(v) for k,v in counter_dict.items()}  # typecast counter_dict to dict
```

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:583]: There is  a `TINY_VALUE` variable we use for avoid division by zero like [here](https://github.com/cleanlab/cleanlab/blob/851831e643943966392285b134bb713ff6b78a75/cleanlab/object_detection/rank.py#L1118) if you can change to that instead

File: tests/test_object_detection.py
Comment (ulya-tkch) [inline tests/test_object_detection.py:977]: Can you add a simple test for `get_average_per_class_confusion_matrix()` to make sure the right values are calculated? Otherwise test lgtm

**Comment by ulya-tkch** [comment]
Hey @Steven-Yiran great work on this so far, thanks for waiting for our functions to be merged in! The logic LGTM (although I commented where it was a bit confusing). Let me know once you are done with the changes and I'll be happy to take a final look :D

File: cleanlab/object_detection/summary.py
Comment (aditya1503) [reply cleanlab/object_detection/summary.py]: added here: https://github.com/Steven-Yiran/cleanlab/blob/8fb478e87af9a5c27d215eb1b025e176c311143f/cleanlab/object_detection/summary.py#L590C1-L590C1

File: tests/test_object_detection.py
Comment (aditya1503) [reply tests/test_object_detection.py]: Added here: https://github.com/Steven-Yiran/cleanlab/blob/8fb478e87af9a5c27d215eb1b025e176c311143f/tests/test_object_detection.py#L990

File: cleanlab/object_detection/summary.py
Comment (aditya1503) [reply cleanlab/object_detection/summary.py]: Added support for class names

**Comment by Steven-Yiran** [comment]
> Hey @Steven-Yiran great work on this so far, thanks for waiting for our functions to be merged in! The logic LGTM (although I commented where it was a bit confusing). Let me know once you are done with the changes and I'll be happy to take a final look :D

Hi @ulya-tkch, I have added all the modifications on my end.

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:528]: I get KeyError when trying to run 
```python
class_names = {"0": "car", "1": "hand", "2": "bowl", "3": "human", "4": "traffic light"}
get_average_per_class_confusion_matrix(labels=labels,predictions=predictions,num_procs=1,class_names=class_names)
```

Reason seems to be avg_metrics is initialized as dict of {class_num: {}} while access in loop is `avg_metrics[class_names[class_num]]["FP"] += fp`.  I would suggest initializing avg_metrics as class_names values instead of keys if class_names is passed in.

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:552]: This is also broken due to issue above. I suggest implementing fix + expending tests to using dictionary of class_names. Otherwise the documentation LGTM after this is fixed, should be good to go.

File: cleanlab/object_detection/summary.py
Comment (aditya1503) [reply cleanlab/object_detection/summary.py:550]: It's changed to be sorted by indices, should be fine now

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:523]:
```suggestion
```

File: cleanlab/object_detection/summary.py
Comment (ulya-tkch) [inline cleanlab/object_detection/summary.py:530]:
```suggestion
   The average confusion matrix provides insights into the model strengths and potential biases.

    Note:  lower TP at certain IoU thresholds does not necessarily imply that everything else is FP, instead it indicates that, at those specific IoU thresholds, the model is not performing as well in terms of correctly identifying class instances. The other metrics (FP and FN) provide additional information about the model's behavior.

    Note: Since we average over many IoU thresholds, 'TP', 'FP', and 'FN' may contain float values representing the average across these thresholds.
```

**Review by ulya-tkch — APPROVED** [review]
Reordered explanation. Otherwise, LGTM! Great work @Steven-Yiran and @aditya1503 on these metrics. They provide a very useful summary for anyone using our object detection code!

**Comment by Steven-Yiran** [comment]
Thanks, @ulya-tkch and @aditya1503 for all the insightful work and guidance! I will follow up with another commit to the Object Detection Tutorial.

**Pull request merged**
